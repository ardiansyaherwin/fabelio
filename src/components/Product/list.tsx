import React from 'react'
import { IProduct } from './index'
import {
  Card,
  CardContent,
  Typography,
  Chip,
  Grid,
} from '@material-ui/core'
import { numberToCurrency } from '../../utils/formatter'
import { createStyles, Theme, makeStyles } from '@material-ui/core/styles'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    tagContainer: {
      'margin': `0px -${theme.spacing(0.5)}px`,
      '& > *': {
        margin: theme.spacing(0.5),
      },
    },
    infoText: {
      fontSize: theme.spacing(1.5),
      fontStyle: 'italic',
      float: 'right',
    },
  }),
)

interface IListProps {
  item: IProduct,
}

export const List = (props: IListProps) => {
  const { item } = props
  const desc = item.description.length > 114 ? item.description.substring(0, 114) + '...' : item.description
  const classes = useStyles()

  return (
    <Card>
      <CardContent>
        <Grid container alignItems='baseline'>
          <Grid item xs={8}>
            <Typography gutterBottom variant='h5' component='h5'>
              {item.name}
            </Typography>
          </Grid>
          <Grid item xs={4}>
            <Typography variant='subtitle1' color='secondary' component='h5' align='right'>
              {numberToCurrency(item.price)}
            </Typography>
          </Grid>
        </Grid>

        <Typography variant='body2' color='textSecondary' component='p' paragraph>
          { desc }
        </Typography>

        <Grid container alignItems='center'>
          <Grid item xs={8} className={classes.tagContainer}>
            {item.furniture_style.map((style: string, index: number) => {
              return <Chip key={index} label={style} color='default' />
            })}
          </Grid>
          <Grid item xs={4}>
            <Typography variant='body2' color='textSecondary' className={classes.infoText} align='right'>
              Delivery time:<br/>{item.delivery_time} day(s)
            </Typography>
          </Grid>
        </Grid>
      </CardContent>
    </Card>
  )
}
