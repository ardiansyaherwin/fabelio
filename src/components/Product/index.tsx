import { List } from './list'

export interface IProduct {
  delivery_time: string
  description: string
  name: string
  price: number
  furniture_style: string[]
  [x: string]: any
}

export {
  List,
}
