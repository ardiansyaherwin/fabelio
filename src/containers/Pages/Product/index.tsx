import React, {useState, useEffect, useRef, useLayoutEffect} from 'react'
import { IProduct } from '@components/Product/index'
import { List } from '@components/Product/list'
import MultipleSelect from '@components/DropdownSelect'
import {
  Container,
  CircularProgress,
  Grid,
  TextField,
  Snackbar,
  AppBar,
  Toolbar,
  Typography,
} from '@material-ui/core'
import MuiAlert from '@material-ui/lab/Alert'
import ProductService from '../../../services/ProductSevices'

interface IProductState {
  errorMessage: string,
  isError: boolean,
  isFetching: boolean,
  products: IProduct[],
  vertical: any,
  horizontal: any,
  furnitureStyleList: string[],
  filteredProducts: IProduct[],
  [listFilter: string]: any,
}

const ProductList = () => {
  const didmountEffect = useRef(true)
  const [productState, setProductState] = useState<IProductState>({
    errorMessage: '',
    isError: false,
    isFetching: true,
    products: [],
    vertical: 'top' as const,
    horizontal: 'center' as const,
    furnitureStyleList: [],
    filteredProducts: [],
    listFilter: {
      nameFilter: '',
      deliveryTimeFilter: '',
      furnitureStyleFilter: '',
    },
  })

  const {
    errorMessage,
    isError,
    isFetching,
    products,
    vertical,
    horizontal,
    furnitureStyleList,
    filteredProducts,
    listFilter,
  } = productState

  const { nameFilter, deliveryTimeFilter, furnitureStyleFilter } = listFilter
  const deliveryTimeList = ['1 week', ' 2 weeks', '1 month', 'more']
  const deliveryTimeListInWeek = [1, 2, 4, 5]
  function Alert(props: any) {
    return <MuiAlert elevation={6} variant='filled' {...props} />
  }

  async function fetchProductList() {
    ProductService().getProductList()
      .then((result: any) => {
        const isSuccess = result.status === 200
        const fetchedProducts = result.data.products

        // for testing purpose
        // products[products.length] = {
        //   delivery_time: '45',
        //   description: 'Woles juga nyampenya',
        //   furniture_style: ['Midcentury'],
        //   name: 'Forbyta Sofa Bed Woles',
        //   price: 8999000,
        // }

        setProductState({
          ...productState,
          errorMessage: result.error ? 'Please try again later' : '',
          isError: isSuccess ? false : true,
          isFetching: false,
          products: isSuccess ? fetchedProducts : null,
          filteredProducts: isSuccess ? fetchedProducts : null,
          furnitureStyleList: result.data.furniture_styles,
        })
        return true
      }).catch(() => {
        setProductState({
          ...productState,
          errorMessage: 'Failed. Please try again later',
          isError: true,
          isFetching: false,
        })
      })
  }

  function filterName(product: IProduct) {
    return product.name.toLowerCase().includes(nameFilter.toLowerCase()) && product
  }

  function filterDeliveryTime(product: IProduct) {
    let currentFilteredProduct: IProduct = null
    const productDeliveryTimeInWeek = Math.ceil(parseInt(product.delivery_time, 0) / 7)

    deliveryTimeFilter.map( (filter: string) => {
      const selectedIndexFilter = deliveryTimeList.indexOf(filter)
      const selectedFilterInWeek = deliveryTimeListInWeek[selectedIndexFilter]
      if (selectedFilterInWeek <= 2) {
        if (productDeliveryTimeInWeek === selectedFilterInWeek) {
          currentFilteredProduct = product
        }
      } else if (selectedFilterInWeek > 2 && selectedFilterInWeek <= 4) {
        if (productDeliveryTimeInWeek > 2 && productDeliveryTimeInWeek <= 4) {
          currentFilteredProduct = product
        }
      } else {
        if (productDeliveryTimeInWeek > 4) {
          currentFilteredProduct = product
        }
      }
    })
    return currentFilteredProduct
  }

  function filterFurniture(product: IProduct) {
    let currentFilteredProduct: IProduct = null
    product.furniture_style.map( (style: string) => {
      if (furnitureStyleFilter.indexOf(style) > -1) {
        currentFilteredProduct = product
      }
      return style
    })
    return currentFilteredProduct
  }

  function filteringList() {
    let filteredList: IProduct[] = products
    if (nameFilter !== '') {
      filteredList = filteredList.filter(filterName)
    }

    if (deliveryTimeFilter.length) {
      filteredList = filteredList.filter(filterDeliveryTime)
    }

    if (furnitureStyleFilter.length) {
      filteredList = filteredList.filter(filterFurniture)
    }

    setProductState({
      ...productState,
      filteredProducts: filteredList,
    })
  }

  function handleClose() {
    setProductState({
      ...productState,
      isError: false,
      errorMessage: '',
    })
  }

  function handleFilterChange(values: any, filter: string) {
    listFilter[filter] = values
    setProductState({
      ...productState,
      listFilter,
    })
  }

  useEffect(() => {
    fetchProductList()
  }, [])

  useLayoutEffect(() => {
    if (didmountEffect.current) {
      didmountEffect.current = false
      return
    }
    filteringList()
  }, [nameFilter, deliveryTimeFilter, furnitureStyleFilter])

  return (
    <div style={{marginTop: 100, paddingTop: 80, paddingBottom: 30}}>
      <Snackbar
        open={isError}
        autoHideDuration={3000}
        onClose={handleClose}
        anchorOrigin={{ vertical, horizontal}}
        key={`${vertical},${horizontal}`}
      >
        <Alert onClose={handleClose} severity='error'>{errorMessage}</Alert>
      </Snackbar>
      <AppBar color='default' position='fixed'>
        <Toolbar>
          <Container maxWidth='md'>
            <Grid container spacing={3} style={{marginBottom: 16}}>
              <Grid item xs={12}>
                <TextField
                  placeholder='Search Furniture'
                  fullWidth
                  margin='normal'
                  value={nameFilter}
                  onChange={(e) => handleFilterChange(e.currentTarget.value, 'nameFilter')}
                />
              </Grid>
              <Grid item xs={6}>
                <MultipleSelect
                  id='furniture-style'
                  label='Furniture Style'
                  items={furnitureStyleList}
                  onChange={(e) => handleFilterChange(e, 'furnitureStyleFilter')}
                />
              </Grid>
              <Grid item xs={6}>
                <MultipleSelect
                  id='delivery-time'
                  label='Delivery Time'
                  items={deliveryTimeList}
                  onChange={(e) => handleFilterChange(e, 'deliveryTimeFilter')}
                />
              </Grid>
            </Grid>
          </Container>
        </Toolbar>
      </AppBar>
      { isFetching ? <CircularProgress /> :
        <Grid container spacing={3} style={{marginBottom: 16}}>
          { filteredProducts.length ?
            filteredProducts.map( (item: IProduct, index: number) => {
              return(
                <Grid key={index} item xs={12} sm={6}>
                  <List item={item} />
                </Grid>
              )
            })
          :
          <Grid item xs={12}>
            <Typography variant='h2' gutterBottom color='textSecondary'>
              No Data Found.
            </Typography>
          </Grid>
        }
        </Grid>
      }
    </div>
  )
}

export default ProductList
