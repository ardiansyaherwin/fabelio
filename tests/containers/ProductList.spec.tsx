import React from 'react'
import renderer from 'react-test-renderer'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import ProductList from '../../src/containers/Pages/Product'
import { List } from '@components/Product'

configure({ adapter: new Adapter() })

describe('product List Screen', () => {
  const setState = jest.fn()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const useStateMock: any = (initState: any) => [initState, setState]
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should render properly', () => {
    const componentRenderer = renderer.create(<ProductList />)
    const tree = componentRenderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should contain List component', () => {
    useStateMock.isFetching = false
    const screen = shallow(<ProductList />)
    jest.spyOn(React, 'useState').mockImplementation(useStateMock)
    expect(screen.find(List).exists())
  })
})
