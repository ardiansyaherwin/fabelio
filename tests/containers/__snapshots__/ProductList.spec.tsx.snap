// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product List Screen should render properly 1`] = `
<div
  style={
    Object {
      "marginTop": 100,
      "paddingBottom": 30,
      "paddingTop": 80,
    }
  }
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorDefault mui-fixed MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <div
        className="MuiContainer-root MuiContainer-maxWidthMd"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
          style={
            Object {
              "marginBottom": 16,
            }
          }
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
            >
              <div
                className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInputBase-input MuiInput-input"
                  disabled={false}
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="Search Furniture"
                  required={false}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <div>
              <div
                className="MuiFormControl-root makeStyles-formControl-199 MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                  data-shrink={false}
                  id="furniture-style-label"
                >
                  Furniture Style
                </label>
                <div
                  className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                  onClick={[Function]}
                >
                  <div
                    aria-haspopup="listbox"
                    aria-labelledby="furniture-style-label furniture-style"
                    className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                    id="furniture-style"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    role="button"
                    tabIndex={0}
                  >
                    <span
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "&#8203;",
                        }
                      }
                    />
                  </div>
                  <input
                    onAnimationStart={[Function]}
                    type="hidden"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    className="MuiSvgIcon-root MuiSelect-icon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <div>
              <div
                className="MuiFormControl-root makeStyles-formControl-199 MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                  data-shrink={false}
                  id="delivery-time-label"
                >
                  Delivery Time
                </label>
                <div
                  className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                  onClick={[Function]}
                >
                  <div
                    aria-haspopup="listbox"
                    aria-labelledby="delivery-time-label delivery-time"
                    className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                    id="delivery-time"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    role="button"
                    tabIndex={0}
                  >
                    <span
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "&#8203;",
                        }
                      }
                    />
                  </div>
                  <input
                    onAnimationStart={[Function]}
                    type="hidden"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    className="MuiSvgIcon-root MuiSelect-icon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>
  <div
    className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style={
      Object {
        "height": 40,
        "width": 40,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </div>
</div>
`;
